content/algorithms/chain_exchange.md

```diff
@@ -0,0 +1,116 @@
+---
+title: "ChainExchange"
+weight: 5
+dashboardWeight: 1
+dashboardState: stable
+dashboardAudit: missing
+dashboardTests: 0
+---
+
+# ChainExchange
+
+{{< hint info >}}
+**Name**: Chain Exchange
+**Protocol ID**: `/fil/chain/xchg/0.0.1`
+{{< /hint >}}
+
+ChainExchange is a simple request/response protocol that allows Filecoin nodes to request ranges of Tipsets and/or Messages from each other. `ChainExchange` is part of a bigger "family" of protocols that live under the umbrella of `ChainSync`. Please refer to the [ChainSync](blockchain#chainsync) section for an overview of the operational boundaries of each of the protocols in ChainSync.
+
+The `Request` message requests a chain segment of a given length by the hash of the highest Tipset in the segment (not necessarily heaviest Tipset of the current chain). For example, if the current height is at 5000, but a node is missing Tipsets between 4500-4700, then the `Request.Head` requested is 4700 and `Request.Length` is 200.
+
+The `Options` allow the requester to specify whether they want to receive block headers of the Tipsets only, the transaction messages included in every block, or both.
+
+The `Response` contains the requested chain segment in reverse iteration order. Each item in the `Chain` array contains either the block headers for that Tipset if the `Blocks` option bit in the request is set, or the messages across all blocks in that Tipset, if the `Messages` bit is set, or both, if both option bits are set.
+
+Each `CompactedMessages` structure contains the BLS and `secp256k1` messages for the corresponding Tipset in unified arrays that encode each message once regardless of how many times it appears in the Tipset's blocks (to reduce the `Response` size). The mapping between message and the blocks into which they are included is encoded in the `BlsIncludes` and `SecpkIncludes` arrays. These arrays are indexed by the block index in the Tipset and contain in that position an array of message indexes that belong to that block. The messages themselves are in the `Bls` and `Secpk` arrays.
+
+If not all Tipsets requested could be fetched, but the `Head` of the chain segment requested was successfully fetched (and potentially more contiguous Tipsets), then this is not considered an error and a `Partial` response code is returned. The node can continue extending the chain from the partial returned segment onwards.
+
+```go
+type Request struct {
```

Comment:
you can embed this struct directly from the code https://github.com/filecoin-project/lotus/blob/master/chain/exchange/protocol.go#L48-L60 using https://github.com/filecoin-project/specs#embed

Example: 
```
{{<embed src="https://github.com/filecoin-project/lotus/blob/master/chain/exchange/protocol.go" lang="go" symbol="Request" title="Exchange Protocol Request">}}
```

This way its easier to keep in sync.

Reply:
good